text/0035-schema-applies-to-unspecified.md

```diff
@@ -0,0 +1,62 @@
+
+## Motivation
+
+**Current behavior**  
+In Cedar 2.x, if an action's `appliesTo.principalTypes` or `appliesTo.resourceTypes` is not given (or if the entire `appliesTo` element is not given), then it's interpreted as an action that applies to all principal types and resource types.
+
+**Challenges**  
+Discussion on [RFC 24](https://github.com/cedar-policy/rfcs/pull/24) highlighted challenges this introduces, primarily the risk of unintentionally specifying that an action applies to all principal types / resource types and related complications it causes for analysis and the experience of policy validation as error messages become more confusing.
+
+**Feature request**  
+We can mitigate the challenges listed above by making following changes:
+
+1. Specified `appliesTo` but omitted `appliesTo.principalTypes` / `appliesTo.resourceTypes` means that the request component is unspecified, i.e., corresponding to the `None` option in the principal and/or resource component of a [`Request`](https://docs.rs/cedar-policy/2.4.0/cedar_policy/struct.Request.html). Omitted `appliesTo.context` means that the context is an empty record (same as the status quo).
+2. Omitted `appliesTo` means that the action cannot be used in a request. It is used exclusively as an action group.
+3. Disallow empty arrays for `appliesTo.principalTypes` / `appliesTo.resourceTypes`.
+4. Disallow an empty `appliesTo` attribute thus requiring at least one of `principalTypes`, `resourceTypes` or `context` to be specified if `appliesTo` is specified.
+
+This feature request is consistent with the semantics for `appliesTo` proposed in [RFC 24](https://github.com/cedar-policy/rfcs/blob/main/text/0024-schema-syntax.md) for the alternate Cedar schema syntax.
+
+## Detailed design
+
+This RFC will require the following changes:
+
+1. The permissive validator must be updated to treat an omitted `appliesTo.principalTypes` / `appliesTo.resourceTypes` as the _unspecified_ entity type, rather than _any_ entity type. The difference is that the _unspecified_ entity type is a particular type that is different from any defined type, whereas _any_ entity type could be the unspecified type or any other type defined in the schema. Note that this change is only required in the _permissive_ validator, because the corresponding change was already made to strict validation as part of [RFC 19](https://github.com/cedar-policy/rfcs/blob/main/text/0019-stricter-validation.md).
+2. The validator must be updated to treat an omitted `appliesTo` attribute as indication that an action does not apply to any principal or resource types.
+3. The schema parser must be updated to reject empty arrays for `appliesTo.principalTypes` / `appliesTo.resourceTypes`.
+4. The schema parser must be updated to reject empty `appliesTo` attributes.
+
+## Drawbacks
+
+This is a breaking change that will alter the meaning of current schemas.
```

Comment:
Can we provide some kind of automated transformation for people using the existing schemas?

Reply:
It's a little tricky, for reasons discussed in my recent commit adding a "Notes on Backwards Compatibility" section. TLDR: if a user has omitted an `appliesTo` field from an action declaration, then they may have intended for the action to apply to any defined entity or the unspecified entity. These two cases were previously conflated, and we'd need user input to decide which they intended.

But aside from that: how would we provide this automated transformation? As a new API in `cedar-policy`, or a standalone executable? How could ensure that users know to run the transformation?

Reply:
> they may have intended for the action to apply to any defined entity or the unspecified entity.

Did we actually implement this previously? We might have suggested in the docs they could apply to any defined entity, but did validation actually consider each entity individually? I didn't think that it did, at least in strict mode; it basically just considered a hypothetical, unknown entity. @john-h-kastner-aws 

If I am right, then whatever the user intended, they were getting validation tantamount to the `unspecified` entity, so we don't really need to distinguish.

Reply:
Yes, the latest released version of Cedar uses a hypothetical, unknown entity.

In many cases, this is the same as validating with an `unspecified` entity (e.g., it is not possible to access an entity's fields without a `has` check). But the `unspecified` type allows more precision, meaning that a policy may fail to validate (with an `ImpossiblePolicy` error) where it previously succeeded.

Reply:
> then they may have intended for the action to apply to any defined entity or the unspecified entity

Although this is possible, the more likely use case is actually reverse one -- they were creating an action group and intended it to not apply to any entity types.

> But aside from that: how would we provide this automated transformation? As a new API in cedar-policy, or a standalone executable? How could ensure that users know to run the transformation?

I don't have an opinion on standalone vs API but I do think this capability would be beneficial for all customers upgrading to latest Cedar. To make them aware, we'd call it out as part of release notes.

Reply:
> Can you give me an example?

Consider this policy
```cedar
permit(principal == User::"alice", action == Action::"action", resource);
```

In the latest released version of Cedar, if `appliesTo` or `appliesTo.principalTypes` are omitted for `Action::"foo"`, then this policy will validate.

After implementing this RFC, the user could choose either to have the action apply to any defined type, or an unspecified entity. In the former case, the policy will continue to validate. In the latter case it will not because [RFC19](https://github.com/cedar-policy/rfcs/blob/main/text/0019-stricter-validation.md) will type `principal == User::"alice"` as `False`.

Reply:
> are omitted for Action::"foo", then this policy will validate.

Did you mean `Action::"action"` rather than `Action::"foo"` ?

> After implementing this RFC, the user could choose either to have the action apply to any defined principal type,

By "any defined principal type" you mean "any entity type", right? The validator cannot know which ones are intended to be used as principals. So if the user intended this behavior they'd have to list all the entity types explicitly, right?

I find it unlikely that any user intends a policy to actually apply to any entity type in a way that would not match treatment of the unspecified entity. It's a misfeature that the current schema treats a missing principal type as "unspecified and/or any other entity type". This RFC fixes that misfeature.

Perhaps we can write an analysis that warns users of policies like this and prompts them to make the appropriate change?

Reply:
Started a fresh thread to discuss the question of whether we should provide translation between schema versions: https://github.com/cedar-policy/rfcs/pull/35#discussion_r1386795521

Also note that several of my statements above turned out to be incorrect 🤦‍♀️ See https://github.com/cedar-policy/rfcs/pull/35#issuecomment-1799533721